File: grafui/Gistogram/src/main/resources/NativeLib.java
package com.tech.nativelib;

public class NativeLib {

    /**
     * Returns a greeting message from native code.
     */
    public native String stringFromJNI();

    /**
     * Builds logistic data.
     */
    public native void buildLogistic(
            String resultFilePath,
            int cityAmount,
            int roadAmount,
            int cargo
    );

    /**
     * Builds logistic data from a file.
     */
    public native void buildLogisticFromFile(
            String dataFilePath,
            String resultFilePath,
            String startCity,
            String endCity,
            int cargo
    );

    /**
     * Builds logistic vertex data.
     */
    public native void buildLogisticVertex(
            String resultFilePath,
            int cityAmount,
            int roadAmount
    );

    /**
     * Builds logistic vertex data from a file.
     */
    public native void buildLogisticFromFileVertex(
            String dataFilePath,
            String resultFilePath,
            String startCity,
            String endCity
    );

    /**
     * Builds undirected logistic data.
     */
    public native void buildLogisticUndirected(
            String resultFilePath,
            int cityAmount,
            int roadAmount,
            int cargo
    );

    /**
     * Builds undirected logistic data from a file.
     */
    public native void buildLogisticFromFileUndirected(
            String dataFilePath,
            String resultFilePath,
            String startCity,
            String endCity,
            int cargo
    );

    /**
     * Builds undirected logistic vertex data.
     */
    public native void buildLogisticVertexUndirected(
            String resultFilePath,
            int cityAmount,
            int roadAmount
    );

    /**
     * Builds undirected logistic vertex data from a file.
     */
    public native void buildLogisticFromFileVertexUndirected(
            String dataFilePath,
            String resultFilePath,
            String startCity,
            String endCity
    );

    /**
     * Builds a histogram based on input parameters.
     */
    public native void buildGistogram(
            String path,
            String option,
            int amount,
            int enableStatistic,
            int filter1,
            int filter2,
            int filter3,
            int filter4,
            int birthYear,
            String cityStartsWith
    );

    /**
     * Builds a histogram from a file.
     */
    public native void buildGistogramFromFile(
            String path,
            String filePath,
            String option,
            int amount,
            int enableStatistic,
            int filter1,
            int filter2,
            int filter3,
            int filter4,
            int birthYear,
            String cityStartsWith
    );

    /**
     * Builds a bag structure with given constraints.
     */
    public native void buildBag(
            String path,
            int amount,
            int maxCapacity,
            int maxVolume
    );

    static {
        String libraryPath = System.getenv("NATIVE_LIB_PATH");
        if (libraryPath == null) {
            libraryPath = "/home/itech/Desktop/University/AdvancedContainers/api/Gistogram/libNativeLib.so";
        }

        try {
            System.load(libraryPath);
            System.out.println("Native library loaded successfully from " + libraryPath);
        } catch (UnsatisfiedLinkError e) {
            System.err.println("Failed to load native library: " + e.getMessage());
            throw new RuntimeException("Could not load native library", e);
        }
    }
}

File: api/Gistogram/src/main/kotlin/nativelib/NativeLib.kt
package com.tech.nativelib

class NativeLib {

    /**
     * Returns a greeting message from native code.
     */
    external fun stringFromJNI(): String

    /**
     * Builds logistic data.
     */
    external fun buildLogistic(
        resultFilePath: String,
        cityAmount: Int,
        roadAmount: Int,
        cargo: Int
    )

    /**
     * Builds logistic data from a file.
     */
    external fun buildLogisticFromFile(
        dataFilePath: String,
        resultFilePath: String,
        startCity: String,
        endCity: String,
        cargo: Int
    )

    /**
     * Builds logistic vertex data.
     */
    external fun buildLogisticVertex(
        resultFilePath: String,
        cityAmount: Int,
        roadAmount: Int
    )

    /**
     * Builds logistic vertex data from a file.
     */
    external fun buildLogisticFromFileVertex(
        dataFilePath: String,
        resultFilePath: String,
        startCity: String,
        endCity: String
    )

    /**
     * Builds undirected logistic data.
     */
    external fun buildLogisticUndirected(
        resultFilePath: String,
        cityAmount: Int,
        roadAmount: Int,
        cargo: Int
    )

    /**
     * Builds undirected logistic data from a file.
     */
    external fun buildLogisticFromFileUndirected(
        dataFilePath: String,
        resultFilePath: String,
        startCity: String,
        endCity: String,
        cargo: Int
    )

    /**
     * Builds undirected logistic vertex data.
     */
    external fun buildLogisticVertexUndirected(
        resultFilePath: String,
        cityAmount: Int,
        roadAmount: Int
    )

    /**
     * Builds undirected logistic vertex data from a file.
     */
    external fun buildLogisticFromFileVertexUndirected(
        dataFilePath: String,
        resultFilePath: String,
        startCity: String,
        endCity: String
    )

    /**
     * Builds a histogram based on input parameters.
     */
    external fun buildGistogram(
        path: String,
        option: String,
        amount: Int,
        enableStatistic: Int,
        filter1: Int,
        filter2: Int,
        filter3: Int,
        filter4: Int,
        birthYear: Int,
        cityStartsWith: String
    )

    /**
     * Builds a histogram from a file.
     */
    external fun buildGistogramFromFile(
        path: String,
        filePath: String,
        option: String,
        amount: Int,
        enableStatistic: Int,
        filter1: Int,
        filter2: Int,
        filter3: Int,
        filter4: Int,
        birthYear: Int,
        cityStartsWith: String
    )

    /**
     * Builds a bag structure with given constraints.
     */
    external fun buildBag(
        path: String,
        amount: Int,
        maxCapacity: Int,
        maxVolume: Int
    )

    companion object {
        private val LIBRARY_PATH = System.getenv("NATIVE_LIB_PATH") ?: "/home/itech/Desktop/University/AdvancedContainers/api/Gistogram/libNativeLib.so"

        init {
            try {
                System.load(LIBRARY_PATH)
                println("Native library loaded successfully from $LIBRARY_PATH")
            } catch (e: UnsatisfiedLinkError) {
                println("Failed to load native library: ${e.message}")
                throw RuntimeException("Could not load native library", e)
            }
        }
    }

}
